lib/ecto/changeset.ex

```diff
@@ -1015,7 +1024,22 @@ defmodule Ecto.Changeset do
           "please use put_#{tag}/4 instead"
   end
 
-  defp put_change(data, changes, errors, valid?, key, value, _type) do
+  defp put_change(data, changes, errors, valid?, key, value, nil) do
+    put_change(data, changes, errors, valid?, key, value)
+  end
+
+  defp put_change(data, changes, errors, valid?, key, value, type) do
+    case Ecto.Type.cast(type, value) do
+      {:ok, ^value} ->
+        put_change(data, changes, errors, valid?, key, value)
+      {:ok, _} ->
+        raise Ecto.CastError, type: type, value: value
```

Comment:
Well, in my humble opinion this should not be a `CastError` and yes an `ArgumentError` or something similar. Since it's not a cast at all.